### PR TITLE
Add documentation about thread and Fiber.scheduler support

### LINKED ISSUE #493: [BUG] Segmentation fault
Hello there,

I am using this gem for a sinatra application and a call to `exec_params` seems to have caused a SIGSEGV.
So far I was unable to reproduce the issue, but given the time I will keep trying.

Attached is an excerpt from the logs (I am running the application in a container)
[logs.txt](https://github.com/ged/ruby-pg/files/10194498/logs.txt)

### PR DESCRIPTION
**Pull request by larskanis**
Fixes #493

@cbandy @ged Would you mind to check+fix my English?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -173,6 +173,38 @@ The following type maps are prefilled with type mappings from the PG::BasicTypeR
 * PG::BasicTypeMapForQueries - a PG::TypeMapByClass prefilled with encoders for common Ruby value classes
 
 
+## Thread support
+
+PG is thread safe in such a way that different threads can use different PG::Connection objects concurrently.
+However it is not safe to access any Pg objects simultaneously from more than one threads.
+So take care to open a new database server connection for every new thread or use a wrapper library like ActiveRecord that manages connections in a thread safe way.
+
+If messages like the following are printed to stderr, you're probably using one connection by several threads:
```
Comment (ged) [inline README.md:182]: > by several threads

Suggest replacing with `from several threads`.

File: README.md [hunk 2]
```diff
@@ -173,6 +173,38 @@ The following type maps are prefilled with type mappings from the PG::BasicTypeR
...
+    message type 0x31 arrived from server while idle
+    message type 0x32 arrived from server while idle
+    message type 0x54 arrived from server while idle
+    message type 0x43 arrived from server while idle
+    message type 0x5a arrived from server while idle
+
+
+## Fiber IO scheduler support
+
+Pg is fully compatible to `Fiber.scheduler` introduced in Ruby-3.0.
```
Comment (ged) [inline README.md:193]: > compatible to

Suggest replacing with `compatible with`

File: README.md [hunk 3]
```diff
@@ -173,6 +173,38 @@ The following type maps are prefilled with type mappings from the PG::BasicTypeR
...
+On Windows support for `Fiber.scheduler` is available on Ruby-3.1 or newer.
+All possibly blocking IO operations are routed through the `Fiber.scheduler` if one is registered for the running thread.
+That is why pg internally uses the asynchronous libpq interface even for synchronous/blocking method calls.
+It also uses Ruby's DNS resolution instead of libpq's builtin functions.
+
+Internally Pg always uses the nonblocking connection mode of libpq.
+It then behaves like running in blocking mode, but ensures, that all blocking IO is handled in Ruby through a possibly registered `Fiber.scheduler`.
```
Comment (ged) [inline README.md:200]: > but ensures, that

No comma necessary.

File: README.md [hunk 4]
```diff
@@ -173,6 +173,38 @@ The following type maps are prefilled with type mappings from the PG::BasicTypeR
...
+When `PG::Connection.setnonblocking(true)` is called then the nonblocking state keeps enabled, but the additional handling of blocking states is disabled, so that the calling program has to take care about blocking states on its own.
```
Comment (ged) [inline README.md:201]: > state keeps enabled

Suggest `state stays enabled`

> has to take care about blocking states

Suggest `has to handle blocking states`

File: README.md [hunk 5]
```diff
@@ -173,6 +173,38 @@ The following type maps are prefilled with type mappings from the PG::BasicTypeR
...
+
+An exception to this rule are the methods for large objects like `PG::Connection#lo_create` and authentication methods using external libraries (like GSSAPI authentication).
+They are not compatible to `Fiber.scheduler`, so that blocking states are not passed to the registered IO scheduler.
```
Comment (ged) [inline README.md:204]: > compatible to

`compatible with`

File: README.md [hunk 6]
```diff
@@ -173,6 +173,38 @@ The following type maps are prefilled with type mappings from the PG::BasicTypeR
 * PG::BasicTypeMapForQueries - a PG::TypeMapByClass prefilled with encoders for common Ruby value classes
 
 
+## Thread support
+
+PG is thread safe in such a way that different threads can use different PG::Connection objects concurrently.
+However it is not safe to access any Pg objects simultaneously from more than one threads.
```
Comment (ged) [inline README.md:179]: > threads

`thread`

File: README.md [hunk 7]
```diff
@@ -173,6 +173,38 @@ The following type maps are prefilled with type mappings from the PG::BasicTypeR
...
+So take care to open a new database server connection for every new thread or use a wrapper library like ActiveRecord that manages connections in a thread safe way.
```
Comment (ged) [inline README.md:180]: > take care to

Not incorrect, but a more typical phrasing would be `make sure to`.

**Review by ged — COMMENTED** [review]
Some English suggestions.

It's so interesting to me that most of these are preposition differences, which is where I typically run into trouble when speaking German. ;)

**Review by cbandy — COMMENTED** [review]
👍🏻

**Comment by larskanis** [comment]
@ged Hi Michael! I pushed the pg-1.4.6 release to rubygems.org. Could you please update the documentation to ruby-pg on http://deveiate.org/code/pg , so that the links to the markdown files work again?

**Pull request merged**